Modix.Services/GuildStats/GuildStatService.cs

```diff
@@ -0,0 +1,169 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Discord;
+using MediatR;
+using Microsoft.Extensions.Caching.Memory;
+using Modix.Data.Models.Core;
+using Modix.Data.Repositories;
+using Modix.Services.Messages.Discord;
+
+namespace Modix.Services.GuildStats
+{
+    public interface IGuildStatService
+    {
+        /// <summary>
+        /// Gets a list of GuildInfoResult objects representing the role distriution for the given guild.
+        /// </summary>
+        /// <param name="guild">The guild to retrieve roles/counts from</param>
+        /// <returns>A list of GuildInfoResult(s), each representing a role in the guild</returns>
+        Task<List<GuildRoleCount>> GetGuildMemberDistributionAsync(IGuild guild);
+
+        /// <summary>
+        /// Returns a mapping of <see cref="GuildUserEntity"/> to a count of the messages they've sent
+        /// </summary>
+        /// <param name="guildId">The guild to count messages for</param>
+        /// <param name="after">How long before now to count messages for</param>
+        Task<IReadOnlyDictionary<GuildUserEntity, int>> GetTopMessageCounts(IGuild guild);
+    }
+
+    public class GuildStatService :
+        INotificationHandler<UserJoined>,
+        INotificationHandler<UserLeft>,
+        INotificationHandler<ChatMessageReceived>,
+        IGuildStatService
+    {
+        private readonly IMemoryCache _cache;
+        private readonly IMessageRepository _messageRepository;
+
+        private readonly MemoryCacheEntryOptions _roleCacheEntryOptions =
+            new MemoryCacheEntryOptions().SetSlidingExpiration(TimeSpan.FromDays(1));
+
+        private readonly MemoryCacheEntryOptions _msgCountCacheEntryOptions =
+            new MemoryCacheEntryOptions().SetSlidingExpiration(TimeSpan.FromMinutes(30));
+
+        public GuildStatService(IMemoryCache cache, IMessageRepository messageRepository)
+        {
+            _cache = cache;
+            _messageRepository = messageRepository;
+        }
+
+        /// <summary>
+        /// Create a unique key object for the cache
+        /// </summary>
+        private object GetKeyForGuild(IGuild guild) => new { guild, Target = "GuildInfo" };
+
+        /// <summary>
+        /// Create a unique key object for the cache
+        /// </summary>
+        private object GetKeyForMsgCounts(IGuild guild) => new { guild, Target = "MessageCounts" };
+
+        /// <summary>
+        /// Clear the cache entry for the given guild
+        /// </summary>
+        public void ClearCacheEntry(IGuild guild)
+        {
+            _cache.Remove(GetKeyForGuild(guild));
+        }
+
+        public Task Handle(ChatMessageReceived notification, CancellationToken cancellationToken)
+        {
+            return Task.CompletedTask;
+        }
```

Comment:
Is this method supposed to be a no-op? If so, do we even need to implement `INotificationHandler<ChatMessageReceived>`?

Reply:
That was in previously but there's not much point in invalidating the cache per-message. Removed.